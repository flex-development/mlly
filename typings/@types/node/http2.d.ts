declare module 'http2' {
  import * as stream from 'node:stream'
  import * as url from 'node:url'

  export interface ClientSessionOptions extends SessionOptions {
    createConnection?(authority: url.URL, option: SessionOptions): stream.Duplex
  }
}
