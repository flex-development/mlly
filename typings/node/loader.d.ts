declare global {
  /**
   * Determines how `url` should be interpreted, retrieved, and parsed.
   * Also in charge of validating import assertions, `context.importAssertions`.
   *
   * @see https://nodejs.org/docs/latest-v16.x/api/all.html#all_esm_loadurl-context-defaultload
   *
   * @async
   *
   * @param {string} url - `file:` url of module
   * @param {LoadHookContext} context - Hook context
   * @param {(LoaderHookFormat | null)?} [context.format] - Module format
   * @param {ImportAssertions} context.importAssertions - Import assertions map
   * @param {LoadHook} defaultLoad - Default Node.js `load` function
   * @return {Promise<LoadHookResult>} Hook result
   */
  declare type LoadHook = (
    url: string,
    context: LoadHookContext,
    defaultLoad: LoadHook
  ) => Promise<LoadHookResult>

  /**
   * {@link LoadHook} context.
   */
  declare interface LoadHookContext {
    /**
     * Module format.
     */
    format?: ResolveHookResult['format']

    /**
     * Import assertions map.
     */
    importAssertions?: ImportAssertions
  }

  /**
   * {@link LoadHook} result.
   */
  declare interface LoadHookResult {
    /**
     * Module format.
     */
    format: LoaderHookFormat

    /**
     * Source code
     */
    source?: ArrayBuffer | Buffer | string | undefined
  }

  /**
   * Valid module formats.
   */
  declare type LoaderHookFormat =
    | 'builtin'
    | 'commonjs'
    | 'dynamic'
    | 'json'
    | 'module'
    | 'wasm'

  /**
   * Returns the resolved file URL for `specifier` and `context.parentURL` and,
   * optionally, its format as a hint to {@link LoadHook}.
   *
   * @see https://nodejs.org/docs/latest-v16.x/api/all.html#all_esm_resolvespecifier-context-defaultresolve
   *
   * @async
   *
   * @param {string} specifier - Module specifier
   * @param {ResolveHookContext} context - Hook context
   * @param {string[]} context.conditions - Import conditions
   * @param {ImportAssertions} context.importAssertions - Import assertions map
   * @param {string} [context.parentURL] - `file:` url of importer
   * @param {ResolveHook} defaultResolve - Node.js default resolver
   * @return {Promise<ResolveHookResult>} Hook result
   */
  declare type ResolveHook = (
    specifier: string,
    context: ResolveHookContext,
    defaultResolve: ResolveHook
  ) => Promise<ResolveHookResult>

  /**
   * {@link ResolveHook} context.
   */
  declare interface ResolveHookContext {
    /**
     * Import conditions.
     */
    conditions: LoaderHookFormat | null

    /**
     * Import assertions map.
     */
    importAssertions: ImportAssertions

    /**
     * `file:` url of importer.
     */
    parentURL?: string
  }

  /**
   * {@link ResolveHook} result.
   */
  declare interface ResolveHookResult {
    /**
     * Module format.
     */
    format?: LoaderHookFormat | null

    /**
     * Absolute url to import target.
     *
     * @example
     *  'file://...'
     */
    url: string
  }
}

export {}
