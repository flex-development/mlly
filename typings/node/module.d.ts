declare module 'module' {
  global {
    /**
     * Filename resolver.
     *
     * @param {string} specifier - Module specifier
     * @param {NodeJS.Module | null} parent - Module that imported `specifier`
     * @return {string} Resolved file URL
     */
    export type ResolveFilename = (
      specifier: string,
      parent: NodeJS.Module | null
    ) => string
  }
}
