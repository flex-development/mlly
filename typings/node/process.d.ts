declare namespace NodeJS {
  interface Process {
    browser?: boolean | null | undefined
  }
}
