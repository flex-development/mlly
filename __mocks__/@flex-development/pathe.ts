/**
 * @file Mocks - @flex-development/pathe
 * @module mocks/flex-development/pathe
 * @see https://github.com/flex-development/pathe
 */

/**
 * [`@flex-development/pathe`][1] module type.
 *
 * [1]: https://github.com/flex-development/pathe
 */
type Actual = typeof import('@flex-development/pathe')

/**
 * `@flex-development/pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('@flex-development/pathe')

export default {
  basename: vi.fn(actual.basename),
  changeExt: vi.fn(actual.changeExt),
  dirname: vi.fn(actual.dirname),
  extname: vi.fn(actual.extname),
  format: vi.fn(actual.format),
  isAbsolute: vi.fn(actual.isAbsolute),
  join: vi.fn(actual.join),
  parse: vi.fn(actual.parse),
  relative: vi.fn(actual.relative),
  removeExt: vi.fn(actual.removeExt),
  resolve: vi.fn(actual.resolve)
}
