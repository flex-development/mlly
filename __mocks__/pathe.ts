/**
 * @file Mocks - pathe
 * @module mocks/pathe
 * @see https://github.com/unjs/pathe
 */

/**
 * [`pathe`][1] module type.
 *
 * [1]: https://github.com/unjs/pathe
 */
type Actual = typeof import('pathe')

/**
 * `pathe` module.
 *
 * @const {Actual} actual
 */
const actual: Actual = await vi.importActual<Actual>('pathe')

export const basename = vi.fn(actual.basename)
export const dirname = vi.fn(actual.dirname)
export const extname = vi.fn(actual.extname)
export const isAbsolute = vi.fn(actual.isAbsolute)
export const join = vi.fn(actual.join)
export const parse = vi.fn(actual.parse)
export const relative = vi.fn(actual.relative)
export const resolve = vi.fn(actual.resolve)
