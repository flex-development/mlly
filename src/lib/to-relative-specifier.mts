/**
 * @file toRelativeSpecifier
 * @module mlly/lib/toRelativeSpecifier
 */

import toUrl from '#lib/to-url'
import type { ModuleId } from '@flex-development/mlly'
import pathe from '@flex-development/pathe'

/**
 * Turn `url` into a *relative specifier*.
 *
 * ::: info
 * The relative specifier will only include a file extension if `specifier`
 * includes a file extension.
 * :::
 *
 * @see {@linkcode ModuleId}
 * @see https://nodejs.org/api/esm.html#terminology
 *
 * @param {ModuleId} url
 *  The `file:` URL to convert
 * @param {ModuleId} parent
 *  Parent module id
 * @return {string}
 *  Relative specifier
 */
function toRelativeSpecifier(url: ModuleId, parent: ModuleId): string {
  /**
   * Relative specifier.
   *
   * @var {string} specifier
   */
  let specifier: string = pathe.relative(
    pathe.fileURLToPath(toUrl(parent)),
    pathe.fileURLToPath(toUrl(url))
  )

  if (/(?:\.\.\/){2,}/.test(specifier)) {
    specifier = specifier.slice(specifier.indexOf(pathe.sep) + 1)
  } else if (specifier.startsWith(pathe.dot.repeat(2))) {
    specifier = specifier.slice(1)
  } else if (specifier) {
    specifier = pathe.dot + pathe.sep + specifier
  } else {
    specifier = pathe.dot
  }

  return specifier
}

export default toRelativeSpecifier
