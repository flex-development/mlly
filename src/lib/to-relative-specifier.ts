/**
 * @file toRelativeSpecifier
 * @module mlly/lib/toRelativeSpecifier
 */

import { fileURLToPath, URL } from 'node:url'
import * as pathe from 'pathe'

/**
 * Converts `specifier` into a relative specifier.
 *
 * **Note**: The returned specifier will only include a file extension if the
 * original specifier includes an extension.
 *
 * @see https://nodejs.org/api/esm.html#terminology
 *
 * @param {URL | string} specifier - Module specifier to convert
 * @param {URL | string} parent - Parent module URL or path to resolve from
 * @return {string} `specifier` as relative specifier
 */
const toRelativeSpecifier = (
  specifier: URL | string,
  parent: URL | string
): string => {
  // convert file url objects to file url strings
  if (parent instanceof URL) parent = parent.href
  if (specifier instanceof URL) specifier = specifier.href

  // convert file url strings to paths
  if (parent.startsWith('file:')) parent = fileURLToPath(parent)
  if (specifier.startsWith('file:')) specifier = fileURLToPath(specifier)

  return pathe
    .relative(pathe.dirname(parent), specifier)
    .replace(/^(\w)/, './$1')
}

export default toRelativeSpecifier
