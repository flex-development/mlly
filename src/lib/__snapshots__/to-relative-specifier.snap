// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/toRelativeSpecifier > should return relative specifier (0) 1`] = `"."`;

exports[`unit:lib/toRelativeSpecifier > should return relative specifier (1) 1`] = `"./cwd"`;

exports[`unit:lib/toRelativeSpecifier > should return relative specifier (2) 1`] = `"../to-relative-specifier"`;

exports[`unit:lib/toRelativeSpecifier > should return relative specifier (3) 1`] = `"."`;
